cilium/CFP-33716-distributed-packet-tracing.md

```diff
@@ -0,0 +1,586 @@
+
+## Goals
+- For tagged packets, add the ID to emitted flow events.
+- For tagged packets, emit datapath debugging events regardless of the debug-verbose setting.
+- Support both IPv4 and IPv6.
+- Add hubble filter for trace ID.
+- Allow users to define which flow IDs they want to monitor for events.
+
+## Non-goals
+- Automatically apply IP option to packets (see Future Milestones).
+
+## User Stories
+- As a platform administrator managing multiple clusters, users are reporting connection timeouts reaching destinations in remote clusters. Platform admin generates traffic with trace ID IP option along the problematic path to identify all flow events associated with that packet at both the source and destination.
+
+![User story](./images/CFP-33716-user-story.png)
+
+- As a production cluster administrator, particular flows from a client are experiencing connection timeouts (e.g. [issue #32472](https://github.com/cilium/cilium/issues/32472). Enabling datapath debugging for all flows leads to event loss and low SNR. To extract dataplane information to root cause the issue, generate traffic with trace ID IP option to emit datapath debug information specific to this flow.
+
+## Proposal
+
+### Overview
+Offer support for packets containing a user-defined IP option field. For tagged incoming flows, handle events for this flow specially; emit datapath debugging events for these packets, include the ID in the event message, and implement hubble filtering by the ID.
+
+### Section 1: Packet Tagging 
+### Packet Tagging Responsibility
+
+This feature does not implement the packet tagging (see Deferred Milestone 1: Tag Packets). Instead, it is the client's responsibility to tag packets via IP option embedding. Therefore, the key assumption for this feature is that tagged packets arrive at ingress points with an `ip_trace_id` already embedded in the IP options.
+
+#### Reasons for Not Implementing Automatic Packet Tagging:
+
+1. **Avoid MTU Concerns:**
+   - Automatically tagging packets can cause them to exceed the MTU, leading to fragmentation and performance issues.
+
+2. **Prevent Performance Loss from GRO:**
+   - GRO engine does not support packets with IP options.
+
+3. **Avoid Packet Loss:**
+   - Some network devices and paths may not support IP options, causing tagged packets in the path to be dropped.
+   - Targets that don't support IP Options:
+      - L4 ELB
+   - Targets that support IP Options:
+      - L4 ILB
+      - Services
+      - Pod-to-pod (including across VPC)
+
+4. **IP Checksum Recalculation:**
+   - When IP options are added or modified, the IP checksum must be recalculated, adding additional processing overhead and complexity.
+
+By making packet tagging the client's responsibility, we ensure that Cilium remains unaffected by these concerns, leaving it to the clients to incorporate these key considerations into their use of this feature and decide how best to utilize it to suit their specific needs.
```

Comment:
It would be good to lift this to be a security section by itself, just so people reading can't miss the context here.